providers/dns/otc/client.go

```diff
@@ -66,3 +75,183 @@ type recordSet struct {
 type recordSetsResponse struct {
 	RecordSets []recordSet `json:"recordsets"`
 }
+
+// Starts a new OTC API Session. Authenticates using userName, password
+// and receives a token to be used in for subsequent requests.
+func (d *DNSProvider) login() error {
+	return d.loginRequest()
+}
+
+func (d *DNSProvider) loginRequest() error {
+	userResp := userResponse{
+		Name:     d.config.UserName,
+		Password: d.config.Password,
+		Domain: nameResponse{
+			Name: d.config.DomainName,
+		},
+	}
+
+	loginResp := loginResponse{
+		Auth: authResponse{
+			Identity: identityResponse{
+				Methods: []string{"password"},
+				Password: passwordResponse{
+					User: userResp,
+				},
+			},
+			Scope: scopeResponse{
+				Project: nameResponse{
+					Name: d.config.ProjectName,
+				},
+			},
+		},
+	}
+
+	body, err := json.Marshal(loginResp)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(http.MethodPost, d.config.IdentityEndpoint, bytes.NewReader(body))
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	client := &http.Client{Timeout: d.config.HTTPClient.Timeout}
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= 400 {
+		return fmt.Errorf("OTC API request failed with HTTP status code %d", resp.StatusCode)
+	}
+
+	d.token = resp.Header.Get("X-Subject-Token")
+
+	if d.token == "" {
+		return fmt.Errorf("unable to get auth token")
+	}
+
+	var endpointResp endpointResponse
+
+	err = json.NewDecoder(resp.Body).Decode(&endpointResp)
+	if err != nil {
+		return err
+	}
+
+	for _, v := range endpointResp.Token.Catalog {
+		if v.Type == "dns" {
+			for _, endpoint := range v.Endpoints {
+				d.baseURL = fmt.Sprintf("%s/v2", endpoint.URL)
+				continue
```

Comment:
This doesn't make a lot of sense. It loops over the catalog, then if it finds the dns entries, ends up setting the last endpoint to `d`, did you mean break here?

Reply:
It's not my code, I just move this section. I will check the algo.